metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -817,27 +817,25 @@ public void run() throws Exception {
                         // succeed; if it does not, that's a fatal error. It is important to do this before
                         // scheduling the record for Raft replication.
                         int recordIndex = 0;
-                        long nextWriteOffset = offsetControl.nextWriteOffset();
+                        long lastOffset = raftClient.prepareAppend(controllerEpoch, records);
+                        long baseOffset = lastOffset - records.size() + 1;
                         for (ApiMessageAndVersion message : records) {
-                            long recordOffset = nextWriteOffset + recordIndex;
+                            long recordOffset = baseOffset + recordIndex;
                             try {
                                 replay(message.message(), Optional.empty(), recordOffset);
                             } catch (Throwable e) {
```

Comment:
Do we need to do any cleanup to RaftClient if we hit a replay exception? Or is that handled by the resign code path?

Reply:
Yeah. It is handle by the controller renounce code. The old code left it to the state machine (controller) to determine if a failure to write/append should result in a leader change. It is probably good to keep the same semantic unless we have a good reason to change it.